File: app/src/main/java/com/redmagic/touchsettings/RootShell.java
package com.redmagic.touchsettings;

import java.io.DataOutputStream;
import java.io.BufferedReader;
import java.io.InputStreamReader;

/** Minimal root helper. Runs commands via `su`. Writing Settings.Global needs root. */
public final class RootShell {

    private RootShell() {}

    public static boolean isRootAvailable() {
        return runOne("id").contains("uid=0");
    }

    /** Runs a single command as root, returns combined stdout. */
    public static String runOne(String command) {
        StringBuilder out = new StringBuilder();
        Process p = null;
        try {
            p = Runtime.getRuntime().exec("su");
            DataOutputStream os = new DataOutputStream(p.getOutputStream());
            os.writeBytes(command + "\n");
            os.writeBytes("exit\n");
            os.flush();
            BufferedReader reader = new BufferedReader(new InputStreamReader(p.getInputStream()));
            String line;
            while ((line = reader.readLine()) != null) {
                out.append(line).append('\n');
            }
            p.waitFor();
        } catch (Exception e) {
            return "ERROR: " + e.getMessage();
        } finally {
            if (p != null) p.destroy();
        }
        return out.toString();
    }

    /** Reads a Settings.Global value as root. Returns "" when unset. */
    public static String getGlobal(String key) {
        String r = runOne("settings get global " + key);
        if (r == null) return "";
        r = r.trim();
        if (r.isEmpty() || r.equals("null") || r.startsWith("ERROR")) return "";
        return r;
    }

    /** Writes a raw Settings.Global value as root (single-quoted to protect & + ,). */
    public static String putGlobal(String key, String value) {
        String safe = value.replace("'", "'\\''");
        return runOne("settings put global " + key + " '" + safe + "'");
    }

    public static int getGlobalInt(String key, int def) {
        String r = getGlobal(key);
        try { return r.isEmpty() ? def : Integer.parseInt(r.replaceAll("[^0-9-]", "")); }
        catch (Exception e) { return def; }
    }
}

File: app/src/main/java/com/mora/gamespace/RootShell.kt
package com.mora.gamespace

import java.io.BufferedReader
import java.io.DataOutputStream
import java.io.InputStreamReader

/**
 * Minimal root helper. Runs commands via `su`. Writing Settings.Global needs root
 * (or WRITE_SECURE_SETTINGS granted through the SELinux/permission script).
 * Ported from the test-mora reference (RootShell.java).
 */
object RootShell {

    fun runOne(command: String): String {
        val out = StringBuilder()
        var p: Process? = null
        try {
            p = Runtime.getRuntime().exec("su")
            val os = DataOutputStream(p.outputStream)
            os.writeBytes(command + "\n")
            os.writeBytes("exit\n")
            os.flush()
            val reader = BufferedReader(InputStreamReader(p.inputStream))
            var line: String? = reader.readLine()
            while (line != null) {
                out.append(line).append('\n')
                line = reader.readLine()
            }
            p.waitFor()
        } catch (e: Exception) {
            return "ERROR: " + e.message
        } finally {
            p?.destroy()
        }
        return out.toString()
    }

    fun isRootAvailable(): Boolean = runOne("id").contains("uid=0")

    /** Reads a Settings.Global value as root. Returns "" when unset. */
    fun getGlobal(key: String): String {
        var r = runOne("settings get global " + key)
        r = r.trim()
        if (r.isEmpty() || r == "null" || r.startsWith("ERROR")) return ""
        return r
    }

    /** Writes a raw Settings.Global value as root (single-quoted to protect & + ,). */
    fun putGlobal(key: String, value: String): String {
        val safe = value.replace("'", "'\\''")
        return runOne("settings put global " + key + " '" + safe + "'")
    }

    fun getGlobalInt(key: String, def: Int): Int {
        val r = getGlobal(key)
        return try {
            if (r.isEmpty()) def else r.replace(Regex("[^0-9-]"), "").toInt()
        } catch (e: Exception) {
            def
        }
    }
}
